cache/manager.go

```diff
@@ -1053,27 +1054,57 @@ func (cm *cacheManager) pruneOnce(ctx context.Context, ch chan client.UsageInfo,
 		}
 	}
 
+	// TODO: pick a better path here
+	dstat, err := disk.GetDiskStat(cm.mountPool.tmpdirRoot)
+	if err != nil {
+		return err
+	}
+
 	return cm.prune(ctx, ch, pruneOpt{
 		filter:       filter,
 		all:          opt.All,
 		checkShared:  check,
 		keepDuration: opt.KeepDuration,
-		keepBytes:    opt.KeepBytes,
+		keepBytes:    calculateKeepBytes(totalSize, dstat, opt),
 		totalSize:    totalSize,
 	})
 }
 
-func (cm *cacheManager) prune(ctx context.Context, ch chan client.UsageInfo, opt pruneOpt) (err error) {
-	var toDelete []*deleteRecord
+func calculateKeepBytes(totalSize int64, dstat disk.DiskStat, opt client.PruneInfo) int64 {
+	// 0 values are special, and means we have no keep cap
+	if opt.MaxStorage == 0 && opt.MinStorage == 0 && opt.Free == 0 {
+		return 0
+	}
+
+	// try and keep as many bytes as we can
+	keepBytes := opt.MaxStorage
 
+	// if we need to free up space, then decrease to that
+	if excess := opt.Free - dstat.Free; excess > 0 {
```

Comment:
At least in Windows it looks like `dstat` can be empty. In that case we should not use the values from it.

Reply:
Previously we just logged errors from that function (not sure if https://github.com/moby/buildkit/pull/5079#issuecomment-2361920982 has been addressed yet). I'm not sure how likely this is but can we make it so that prune only errors if `GetDiskStat` fails and `Free` is set. For other deletions I think it could just continue even if the `GetDiskStat` doesn't give good result.

Reply:
Ah I see, I've adjusted to do this now.